### PR TITLE
Refactor code related to labels & contacts uniqueness

### PR DESCRIPTION
**Pull request by asiia-trilitech**
## Proposed changes

Prep for the following task: [Task link](https://app.asana.com/0/1205770721172203/1205770721303329/f)

## Types of changes

- [ ] Bugfix
- [ ] New feature
- [x] Refactor
- [ ] Breaking change
- [ ] UI fix

## Checklist

- [x] Tests that prove my fix is effective or that my feature works have been added
- [x] Documentation has been added (if appropriate)
- [ ] Screenshots are added (if any UI changes have been made)
- [x] All TODOs have a corresponding task created (and the link is attached to it)

### CODE REVIEW

File: src/components/ContactModal.test.tsx [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+import { contact1, contact2 } from "../mocks/contacts";
+import { mockImplicitAddress } from "../mocks/factories";
+import { fireEvent, render, screen, waitFor } from "../mocks/testUtils";
+import { contactsActions } from "../utils/redux/slices/contactsSlice";
+import { store } from "../utils/redux/store";
+
+describe("UpsertContactModal", () => {
+  describe("on adding contact", () => {
+    describe.each([
+      { testCase: "new contact", renderComponent: () => render(<UpsertContactModal />) },
```
Comment (serjonya-trili) [inline src/components/ContactModal.test.tsx:11]: you don't have to use `render` here, you can just leave the element itself and use render in the test case

File: src/components/ContactModal.test.tsx [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+      it("adds contact to address book", async () => {
+        store.dispatch(contactsActions.upsert(contact2));
+        renderComponent();
+
+        // Set name
+        fireEvent.change(screen.getByLabelText("Name"), {
+          target: { value: "Test Contact" },
+        });
+        fireEvent.blur(screen.getByLabelText("Name"));
```
Comment (serjonya-trili) [inline src/components/ContactModal.test.tsx:77]: do you need the `blur` event here? if not maybe you could use `userEvent` instead for writing the text into the input?

File: src/components/ContactModal.test.tsx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+      // Update name
+      fireEvent.change(screen.getByLabelText("Name"), {
+        target: { value: "Updated Name" },
+      });
+      fireEvent.blur(screen.getByLabelText("Name"));
+
+      // Submit
+      await waitFor(() => expect(screen.getByTestId("confirmation-button")).toBeEnabled());
+      fireEvent.click(screen.getByTestId("confirmation-button"));
```
Comment (serjonya-trili) [inline src/components/ContactModal.test.tsx:220]: again, please prefer `userEvent` to `fireEvent`

File: src/components/ContactModal.tsx [hunk 4]
```diff
@@ -20,27 +20,30 @@ import { CopyableAddress } from "./CopyableText";
...
-  const getAccount = useGetOwnedAccountSafe();
   const { isOpen, onClose } = useContext(DynamicModalContext);
 
-  const onSubmitContact = (newContact: Contact) => {
-    if (getAccount(newContact.pkh)) {
-      return;
-    }
-    dispatch(contactsActions.upsert(newContact));
+  // When editing existing contact, its name & pkh are known.
+  const isEdit = contact !== undefined && contact.pkh != "" && contact.name !== "";
```
Comment (serjonya-trili) [inline src/components/ContactModal.tsx:43]: so, if contact is not undefined then why we check that pkh and name are non-empty? they must be, don't they? also, please don't use `!=`, use `!==` instead

File: src/components/ContactModal.tsx [hunk 5]
```diff
@@ -110,29 +88,39 @@ export const UpsertContactModal: FC<{
...
+            )}
           </FormControl>
           <FormControl isInvalid={!!errors.pkh} marginY={5}>
             <FormLabel>Address</FormLabel>
             <Input
               type="text"
               {...register("pkh", {
                 required: "Address is required",
-                validate: validatePkh,
+                validate: isEdit ? () => true : validatePkh,
```
Comment (serjonya-trili) [inline src/components/ContactModal.tsx:101]: well, it's an optimisation, but not very needed tbh. we won't gain much performance gain. it's fine to always validate it

File: src/utils/hooks/contactsHooks.test.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      { type: "ledger", pkh: mockLedgerAccount(0).address.pkh },
+      { type: "social", pkh: mockSocialAccount(1).address.pkh },
+      { type: "secret_key", pkh: mockSecretKeyAccount(2).address.pkh },
+      { type: "mnemonic", pkh: mockMnemonicAccount(3).address.pkh },
+      { type: "multisig", pkh: mockMultisigAccount(0).address.pkh },
+    ])("fails for address used for $type account", ({ pkh }) => {
+      store.dispatch(contactsActions.reset());
+      store.dispatch(accountsSlice.actions.addAccount(mockLedgerAccount(0)));
+      store.dispatch(accountsSlice.actions.addAccount(mockSocialAccount(1)));
+      store.dispatch(accountsSlice.actions.addAccount(mockSecretKeyAccount(2)));
```
Comment (serjonya-trili) [inline src/utils/hooks/contactsHooks.test.ts:133]: there is a neat test helper called just `addAccount` which should shorten all of these

File: src/utils/hooks/contactsHooks.ts [hunk 7]
```diff
@@ -1,21 +1,35 @@
...
-  };
+  return (pkh: string) => pkh in contacts;
 };
 
 export const useGetContactName = () => {
   const contacts = useContacts();
   return (pkh: string) => (pkh in contacts ? contacts[pkh].name : undefined);
 };
+
+export const useValidatePkh = () => {
```
Comment (serjonya-trili) [inline src/utils/hooks/contactsHooks.ts:22]: it's better to be split into several validators. for example, if we use it in the address autocomplete as a destination for a tez transfer, why do we care if it exists in our accounts?
or, if it's intended to be used just for contacts, then let's call it `useValidateNewContactPkh` or something

File: src/components/ContactModal.tsx (hunk 4 above)
Comment (asiia-trilitech) [reply src/components/ContactModal.tsx:43]: The check is needed here at least for the name. 

There are 3 existing use cases for this modal. 

Creating new contact:
- undefined contact (creating from scratch)
- contact with pkh and empty name (adding used address to contacts)

Editing existing contact:
- contact with both pkh and non-empty name

File: src/components/ContactModal.tsx (hunk 5 above)
Comment (asiia-trilitech) [reply src/components/ContactModal.tsx:101]: Without this fix, the check fails for edit case - cause pkh already exists in contacts 🙂

File: src/components/ContactModal.tsx (hunk 4 above)
Comment (serjonya-trili) [reply src/components/ContactModal.tsx:43]: what about `const isEdit = contact?.pkh && contact?.name` then?

Comment (asiia-trilitech) [reply src/components/ContactModal.tsx:43]: Maybe I got something wrong, but it produced the wrong type
`(contact?.pkh && contact.name) ? true : false` worked though 🙂

Comment (serjonya-trili) [reply src/components/ContactModal.tsx:43]: then `const isEdit = !!(contact?.pkh && contact.name)` please

File: src/components/ContactModal.test.tsx (hunk 2 above)
Comment (asiia-trilitech) [reply src/components/ContactModal.test.tsx]: It doesn't work with just `user.type`. For some reason `user.type` doesn't trigger whatever `blur` triggers.
But I managed to make it work with clicking outside of the text input

File: src/utils/hooks/contactsHooks.test.ts (hunk 6 above)
Comment (asiia-trilitech) [reply src/utils/hooks/contactsHooks.test.ts]: Good one, fixed

File: src/components/ContactModal.test.tsx (hunk 2 above)
Comment (serjonya-trili) [reply src/components/ContactModal.test.tsx]: if you need the blur even then please stick to the `fireEvent`

File: src/utils/hooks/contactsHooks.ts (hunk 7 above)
Comment (asiia-trilitech) [reply src/utils/hooks/contactsHooks.ts]: I think we only need a separate check for the contracts - renamed

File: src/utils/hooks/contactsHooks.ts [hunk 8]
```diff
@@ -1,21 +1,35 @@
...
+export const useValidateNewContactPkh = () => {
+  const addressExistsInContacts = useAddressExistsInContacts();
+  const getAccount = useGetOwnedAccountSafe();
+
+  return (pkh: string) => {
+    if (!isAddressValid(pkh)) {
+      return "Invalid address";
+    }
+    if (getAccount(pkh)) {
+      return "Address already used in accounts";
```
Comment (serjonya-trili) [inline src/utils/hooks/contactsHooks.ts:31]: Address *is* ...
same below please

**Review by serjonya-trili — APPROVED** [review]

File: src/utils/hooks/contactsHooks.ts (hunk 8 above)
Comment (asiia-trilitech) [reply src/utils/hooks/contactsHooks.ts]: Thank you, fixed

**Pull request merged**